### DISCUSSION TITLE
Issues with FileInput and useController when attempting to extend Smart Form Component example

### DISCUSSION THREAD

**Discussion by ajzeigert** [react-hook-form/react-hook-form#8436] [Q&A]
Hi, I'm attempting to extend the (very cool) Smart Form Component example in the Advanced section of the docs. 

I've created a handful of reusable form components that use `useController`.

I've also created a conditional section in my form that relies on `watch`. My problem is when I use `input` of `type=file` in a conditional statement. When I submit the form, the file input elements within the conditional section are (visually) reset, although the form data seems to be retained. 

I have attempted to re-create the problem in this CSB:

https://codesandbox.io/s/react-hook-form-js-forked-uj83tm

In the above example, you can: 

1) Choose a file in the `unconditionalFileInput` and submit, see that the data is logged and the input element retains the file name
2) Check the `condition` box, choose files in `conditionalFileInput1` and `conditionalFileInput2`, submit the data and see that the data is logged, but the input elements revert to their default state

My best guess is that it has something to do with my implementation and the fact that my conditional statements are wrapped in a fragment as these sections contain multiple file inputs. Tested in Chrome on macOS 11.6.6.

Any help would be appreciated, thanks in advance.

**Comment by ajzeigert** [comment]
After some further testing, I have discovered the following. Using fragments, as in the following example, exhibits the unwanted visual reset mentioned above:

```
{showConditionalMultipleInput && (
    <>
          <FileInput name="conditionalFileInput1" />
          <FileInput name="conditionalFileInput2" />
    </>
)}
```

However, using multiple conditional statements, as in the following example, works as expected:

```
{showConditionalSingleInput && <FileInput name="conditionalFileInput1" />}
{showConditionalSingleInput && <FileInput name="conditionalFileInput2" />}
```

I have updated the example CSB to include both methods. 

https://codesandbox.io/s/react-hook-form-js-forked-uj83tm

**Comment by ajzeigert** [comment]
After further research, it appears that the code I added to the smart form wasn't working as designed (by me). The original smart form implementation seems to have been intended to register only immediate children as form components. However, my plan required some conditional nested sections and the use of fieldsets, so I needed to recursively register children if they were found. My original implementation did not take into account that React Fragments would not have a `type` property, and therefore I was passing `undefined` as the first argument to `createElement`, which seemed to break the chain somewhere.

I ended up using ReactIs to detect if an element is a Fragment, and then recursively rendering its children, e.g.:

```
...
} else if (ReactIs.isFragment(child) && Array.isArray(child?.props?.children)) {
        return renderChildren(child.props.children);
}
...
```

This allows the conditionals wrapped in a Fragment to render and register correctly. However, wrapping conditional in `fieldset` still results in a seemingly detached visual state, despite the form data still containing to chosen file. I've updated the CSB to include the things that are working and the things that aren't.

https://codesandbox.io/s/react-hook-form-js-forked-uj83tm

Anyway, my current fix is to replace all of my `fieldset` wrappers with Fragments for now, or remove them altogether. I'd still like to know if there's a problem in my implementation that is making `fieldset`s bork the registration somehow.

**Reply by Upinwv** [reply]
`JI`

**Comment by Upinwv** [comment]
https://www.cisco.com/c/en/us/solutions/enterprise-networks/promotions-free-trials.html#sd-wan-routing
